src/java.base/share/classes/javax/crypto/KEM.java

```diff
@@ -0,0 +1,631 @@
+         *          {@code from > to}, or {@code to > secretSize()}
+         * @throws NullPointerException if {@code encapsulation} or
+         *          {@code algorithm} is {@code null}
+         * @throws UnsupportedOperationException if the combination of
+         *          {@code from}, {@code to}, and {@code algorithm}
+         *          is not supported by the decapsulator
+         */
+        public SecretKey decapsulate(byte[] encapsulation,
+                int from, int to, String algorithm)
+                throws DecapsulateException {
+            Objects.checkFromToIndex(from, to, secretSize());
+            var result = d.engineDecapsulate(
+                    Objects.requireNonNull(encapsulation, "null encapsulation"),
+                    from, to,
+                    Objects.requireNonNull(algorithm, "null algorithm"));
+            assert result != null : "null engineDecapsulate result";
+            return result;
+        }
+
+        /**
+         * Returns the size of the shared secret.
+         *
+         * @return the size of the shared secret
+         */
+        public int secretSize() {
+            int result = d.engineSecretSize();
+            assert result >= 0 && result != Integer.MAX_VALUE
+                    : "invalid engineSecretSize result";
+            return result;
+        }
+
+        /**
+         * Returns the size of the key encapsulation message.
+         *
+         * @return the size of the key encapsulation message
+         */
+        public int encapsulationSize() {
+            int result = d.engineEncapsulationSize();
+            assert result >= 0 && result != Integer.MAX_VALUE
+                    : "invalid engineEncapsulationSize result";
+            return result;
+        }
+    }
+
+    /**
+     * An encapsulator, generated by {@link #newEncapsulator}.
+     */
+    public static final class Encapsulator {
+
+        private final KEMSpi.EncapsulatorSpi e;
+        private final Provider p;
+
+        private Encapsulator(KEMSpi.EncapsulatorSpi e, Provider p) {
+            assert e != null;
+            assert p != null;
+            this.e = e;
+            this.p = p;
+        }
+
+        /**
+         * Returns the provider.
+         *
+         * @return thr provider
+         */
+        public Provider provider() {
+            return p;
+        }
+
+        /**
+         * The key encapsulation function.
+         * <p>
+         * This method is equivalent to
+         * {@code encapsulate(0, secretSize(), "Generic")}. This combination
+         * of arguments should be supported by every implementation.
+         *
+         * @return an {@link KEM.Encapsulated} object containing the full
+         *          shared secret as a key with algorithm being "Generic",
+         *          the key encapsulation message, and optional parameters.
+         *          The return value should not be {@code null}.
```

Comment:
"should" means it *could* still return null. I assume that is not what we want. Although I would be more inclined to only specify cases where null may be returned, and if it isn't mentioned, then it should be implied that null is not a legal return value. If in doubt, perhaps check with Joe/CCC for advice when you file the CSR.

This general comment applies to the other return types in this API where you say "not null". I think you can omit those.